src/main/drivers/flash_w25q128fv.c

```diff
@@ -0,0 +1,451 @@
+}
+
+static uint8_t w25q128fv_readRegister(flashDeviceIO_t *io, uint8_t command)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+
+    uint8_t in[1];
+    quadSpiReceive1LINE(quadSpi, command, 0, in, W25Q128FV_STATUS_REGISTER_BITS / 8);
+
+    return in[0];
+}
+
+static void w25q128fv_writeRegister(flashDeviceIO_t *io, uint8_t command, uint8_t data)
+{
+    QUADSPI_TypeDef *quadSpi = io->handle.quadSpi;
+
+   quadSpiTransmit1LINE(quadSpi, command, 0, &data, W25Q128FV_STATUS_REGISTER_BITS / 8);
+}
+
+static void w25q128fv_deviceReset(flashDevice_t *fdevice)
+{
+    flashDeviceIO_t *io = &fdevice->io;
+
+    w25q128fv_waitForReady(fdevice);
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_ENABLE_RESET);
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_RESET_DEVICE);
+
+    w25q128fv_setTimeout(fdevice, W25Q128FV_TIMEOUT_RESET_MS);
+    w25q128fv_waitForTimeout(fdevice);
+
+    w25q128fv_waitForReady(fdevice);
+
+#ifdef DISABLE_NONVOLATILE_QE_MODE // Use this if you encounter a chip with it's QE bit enabled when it shouldn't be.
+    w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_WRITE_ENABLE);
+    w25q128fv_writeRegister(io, W25Q128FV_INSTRUCTION_WRITE_STATUS2_REG, 0x00);
+#endif
+
+
+#if defined(USE_FLASH_WRITES_USING_4LINES) || defined(USE_FLASH_READS_USING_4LINES)
+    uint8_t registerValue;
+    uint8_t newValue;
+    registerValue = w25q128fv_readRegister(io, W25Q128FV_INSTRUCTION_READ_STATUS2_REG);
+
+    //
+    // WARNING: DO NOT ENABLE QE bit if IO2/IO3 are connected to GND or VCC.
+    //
+    // See datasheet https://www.winbond.com/resource-files/w25q128fv%20rev.m%2005132016%20kms.pdf
+    // W25Q128FV - Revision M - 7.1.10 Quad Enable
+    //
+    // There is no such warning for the W25Q128JV in the same documentation section
+    // See datasheet https://www.winbond.com/resource-files/w25q128jv%20revg%2004082019%20plus.pdf
+    // W25Q128JV - Revision G - 7.1.4 Quad Enable
+    //
+
+    if ((registerValue & W25Q128FV_SR2_BIT_QUAD_ENABLE) == 0) {
+        // Enable QUADSPI mode.
+        registerValue = w25q128fv_readRegister(io, W25Q128FV_INSTRUCTION_READ_STATUS2_REG);
+
+        newValue = registerValue;
+        newValue |= W25Q128FV_SR2_BIT_QUAD_ENABLE;
+
+        //w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_WRITE_ENABLE);
+        w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_VOLATILE_WRITE_ENABLE);
+        w25q128fv_writeRegister(io, W25Q128FV_INSTRUCTION_WRITE_STATUS2_REG, newValue);
+    }
+
+    registerValue = w25q128fv_readRegister(io, W25Q128FV_INSTRUCTION_READ_STATUS2_REG);
+    if (registerValue & W25Q128FV_SR2_BIT_QUAD_ENABLE) {
+        // Don't enable QPI mode, otherwise *all* commands and addresses would have to be sent using 4 wires.
+        // If QPI mode disabled commands can still be sent using 1,2 or 4 wires.
+        // w25q128fv_performOneByteCommand(io, W25Q128FV_INSTRUCTION_ENTER_QPI_MODE);
```

Comment:
Seems to be missing code here. Is this complete?

Reply:
yes, it's complete.  It served as a note on how to/not to use the chip and what effect the QE bit would have to would-be readers at the time of implementation.  I can delete the block if you like though, just let me know.

Reply:
block removed.